documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+    /// <param name="value1">The first quaternion.</param>
+    /// <param name="value2">The second quaternion.</param>
+    /// <returns>The quaternion containing the values that result from subtracting each element in <paramref name="value2" /> from its corresponding element in <paramref name="value1" />.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Subtract<T>(Quaternion<T> value1, Quaternion<T> value2)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+    
+    /// <summary>Calculates the length of the quaternion.</summary>
+    /// <returns>The computed length of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T Length<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Calculates the squared length of the quaternion.</summary>
+    /// <returns>The length squared of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T LengthSquared<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+}
+```
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2F
+- Box2I
+- Box3F
+- Box3I
+- CircleF
+- CircleI
+- CubeF
+- CubeI
+- PlaneF
+- PlaneI
+- Ray2F
+- Ray2I
+- Ray3F
+- Ray3I
+- RectangleF
+- RectangleI
+- SphereF
+- SphereI
+
+The Box structs are defined by Min and Max Vectors, while Rectangle and Cube are defined by Origin and Size Vectors. These Types **must** be implicitly castable between each other.
+
+Each type **must** include the following:
+- Intersect functions with both another instance of the type and a point
+- GetDistanceToNearest(Point,Edge,etc) functions **must** be available for a given point
+- For all but the rays and planes, GetInflated function that takes a point and returns the scaled object that is closest to the original and contains the given point
+- Include Scale and Translation transformation functions
+- For Planes and Rays, Normalize functions
+- For Planes include the following static functions
+  - CreateFromVerticies
+  - Dot
+    - with a Vector4
+  - DotCoordinate and DotNormal
+    - with a Vector3
+  - Transform
+    - With a Matrix4x4 or Quaternion, if relevant
+
+# Vector Interface Definitions
+
+The following interfaces are defined for Vectors:
+
+```cs
+public interface IVector
+{
+    /// <summary>
+    /// The type of scalars in this vector type.
+    /// </summary>
+    protected static abstract Type ScalarType { get; }
+    
+    /// <summary>
+    /// The amount of scalars in this vector type.
+    /// </summary>
+    protected internal static abstract int Count { get; }
+}
+
+public interface IVectorEquatable<TVector, TScalar>
+    : IEquatable<TVector>
+    where TVector : IVector<TVector, TScalar>
+    where TScalar : INumberBase<TScalar>
+{
+    public TVector ScalarsEqual(TVector other);
```

Comment:
Is this returning `AllBitsSet` for match and `Zero` for no match?

If so, do the other types need to expose this concept explicitly? Does there then need to be some `IndexOfFirstMatch`, `IndexOfLastMatch`, etc?

Should this just be called `Equals` (like `Vector<T>` and `Vector128<T>`) or maybe `CompareEqual`?

Reply:
What also, about `IVectorComparable` for the other comparison operators?